code/game/machinery/doors/firedoor.dm

```diff
@@ -252,32 +145,253 @@
 		if(CLOSED)
 			nextstate = null
 			close()
-	return
-
-
-/obj/machinery/door/firedoor/border_only
+			
+/obj/machinery/door/firedoor/proc/deconstruct(disassembled = TRUE)
+	if(can_deconstruct)
+		var/obj/structure/firelock_frame/F = new assembly_type(get_turf(src))
+		if(disassembled)
+			F.constructionStep = CONSTRUCTION_PANEL_OPEN
+		else
+			F.constructionStep = CONSTRUCTION_WIRES_EXPOSED
+		F.update_icon()
+	qdel(src)
+	
+/obj/machinery/door/firedoor/border_only	
 	icon = 'icons/obj/doors/edge_Doorfire.dmi'
 	flags = ON_BORDER
 
-/obj/machinery/door/firedoor/border_only/CanPass(atom/movable/mover, turf/target, height=0, air_group=0)
+/obj/machinery/door/firedoor/border_only/CanPass(atom/movable/mover, turf/target, height=0)
 	if(istype(mover) && mover.checkpass(PASSGLASS))
 		return 1
 	if(get_dir(loc, target) == dir) //Make sure looking at appropriate border
-		if(air_group) return 0
 		return !density
 	else
 		return 1
 
-/obj/machinery/door/firedoor/border_only/CheckExit(atom/movable/mover as mob|obj, turf/target as turf)
+/obj/machinery/door/firedoor/border_only/CheckExit(atom/movable/mover as mob|obj, turf/target)
 	if(istype(mover) && mover.checkpass(PASSGLASS))
 		return 1
 	if(get_dir(loc, target) == dir)
 		return !density
 	else
 		return 1
 
-/obj/machinery/door/firedoor/border_only/CanAtmosPass(var/turf/T)
+/obj/machinery/door/firedoor/border_only/CanAtmosPass(turf/T)
 	if(get_dir(loc, T) == dir)
 		return !density
 	else
-		return 1
\ No newline at end of file
+		return 1
+
+/obj/machinery/door/firedoor/heavy
+	name = "heavy firelock"
+	icon = 'icons/obj/doors/Doorfire.dmi'
+	glass = 0
+	opacity = 1
+	assembly_type = /obj/structure/firelock_frame/heavy
+	
+/obj/machinery/door/firedoor/heavy/ex_act(severity)
+	switch(severity)

```

Comment:
I kept the switch for consistency's sake (as most ex_acts use it) but this can also just be a singular if check.